File: lecture0-30/lecture012/mazePath2.java
import java.util.Scanner;
import java.util.ArrayList;

public class mazePath2 {
    public static Scanner scn = new Scanner(System.in);

    public static void main(String[] args) {
        solve();
    }

    public static void solve() {
        System.out.println(mazePath(0, 0, 2, 2)); 
    }

    public static ArrayList<String> mazePath(int sr, int sc, int er, int ec) {
        if (sr == er && sc == ec) {
            ArrayList<String> base = new ArrayList<>();
            base.add("");
            return base;
        }
        ArrayList<String> ans = new ArrayList<>();

        if (sr + 1 <= er) {
            ArrayList<String> vertical = mazePath(sr + 1, sc, er, ec);
            for (String s : vertical) {
                ans.add("V" + s);
            }
        }
        if (sc + 1 <= ec) {
            ArrayList<String> horizontal = mazePath(sr , sc+1, er, ec);
            for (String s : horizontal) {
                ans.add("H" + s);
            }
        }
        if (sr + 1 <= er && sc + 1 <= ec) {
            ArrayList<String> diagonal = mazePath(sr + 1, sc+1, er, ec);
            for (String s : diagonal) {
                ans.add("D" + s);
            }
        }
        return ans;
    }
}

File: lecture015/mazePath2.cpp
#include <iostream>
#include <string>

using namespace std;
int mazePath(int sr, int sc, int er, int ec, string ans)
{
    if (sr == er && sc == ec)
    {
        cout << ans << " ";
        return 1;
    }
    int count = 0;
    if (sr + 1 <= er)
    {
        count += mazePath(sr + 1, sc, er, ec, ans + "V");
    }
    if (sc + 1 <= ec)
    {
        count += mazePath(sr, sc + 1, er, ec, ans + "H");
    }
    if (sc + 1 <= ec && sr + 1 <= er)
    {
        count += mazePath(sr + 1, sc + 1, er, ec, ans + "D");
    }
    return count;
}
int main(int args, char **argv)
{

    cout << mazePath(0, 0, 2, 2, "") << endl;
}